File: foreign/typescript/src/schema-codecs.ts
import { Ajv2020, type AnySchema, type ValidateFunction } from "ajv/dist/2020.js"
import avro, { type Type as AvroType } from "avsc"
import protobuf from "protobufjs"
import "protobufjs/ext/descriptor.js"

import { CodecError, InvalidError } from "./client/errors.js"
import { toNodeBuffer } from "./iggy/apache-iggy.js"
import type { Codec } from "./stream/codecs.js"
import type { SchemaDef } from "./wire/control.js"

export type CompiledSchemaKind = "avro" | "protobuf" | "jsonSchema"

type CompiledDefinition =
  | { readonly kind: "avro"; readonly schema: AvroType }
  | { readonly kind: "protobuf"; readonly message: protobuf.Type }
  | { readonly kind: "jsonSchema"; readonly validate: ValidateFunction }

function codecError(message: string, operation: string, cause?: unknown): CodecError {
  return new CodecError(message, "schema", operation, cause === undefined ? undefined : { cause })
}

function parseJson(text: string, message: string): unknown {
  try {
    return JSON.parse(text) as unknown
  } catch (cause) {
    throw new InvalidError(message, undefined, { cause })
  }
}

function lowerDecodedValue(value: unknown): unknown {
  if (value instanceof Uint8Array) return Array.from(value)
  if (Array.isArray(value)) return value.map(lowerDecodedValue)
  if (value !== null && typeof value === "object") {
    return Object.fromEntries(
      Object.entries(value).map(([key, nested]) => [key, lowerDecodedValue(nested)])
    )
  }
  return value
}

export class CompiledSchema {
  readonly kind: CompiledSchemaKind

  private constructor(private readonly definition: CompiledDefinition) {
    this.kind = definition.kind
  }

  static compile(schema: SchemaDef): CompiledSchema {
    switch (schema.source.kind) {
      case "avro": {
        const source = parseJson(
          schema.source.schema,
          `schema ${String(schema.id)}: Avro schema is not valid JSON`
        )
        try {
          return new CompiledSchema({ kind: "avro", schema: avro.Type.forSchema(source as never) })
        } catch (cause) {
          throw new InvalidError(
            `schema ${String(schema.id)}: Avro schema does not compile`,
            undefined,
            { cause }
          )
        }
      }
      case "protobuf": {
        let root: protobuf.Root
        try {
          root = protobuf.Root.fromDescriptor(schema.source.descriptorSet)
        } catch (cause) {
          throw new InvalidError(
            `schema ${String(schema.id)}: Protobuf descriptor set does not decode`,
            undefined,
            { cause }
          )
        }
        const messageType = schema.source.messageType.replace(/^\./, "")
        try {
          return new CompiledSchema({
            kind: "protobuf",
            message: root.lookupType(messageType)
          })
        } catch (cause) {
          throw new InvalidError(
            `schema ${String(schema.id)}: descriptor set has no message \`${schema.source.messageType}\``,
            undefined,
            { cause }
          )
        }
      }
      case "jsonSchema": {
        const source = parseJson(
          schema.source.schema,
          `schema ${String(schema.id)}: JSON Schema is not valid JSON`
        )
        try {
          const ajv = new Ajv2020({ strict: true })
          return new CompiledSchema({
            kind: "jsonSchema",
            validate: ajv.compile(source as AnySchema)
          })
        } catch (cause) {
          throw new InvalidError(
            `schema ${String(schema.id)}: JSON Schema does not compile`,
            undefined,
            { cause }
          )
        }
      }
      case "unknown":
        throw new InvalidError(`schema ${String(schema.id)}: unknown schema source kind`)
    }
  }

  validate(payload: Uint8Array): boolean {
    try {
      this.decode(payload)
      return true
    } catch {
      return false
    }
  }

  validateValue(value: unknown): boolean {
    return this.definition.kind === "jsonSchema" && this.definition.validate(value)
  }

  decode(payload: Uint8Array): unknown {
    switch (this.definition.kind) {
      case "avro":
        try {
          return lowerDecodedValue(this.definition.schema.fromBuffer(toNodeBuffer(payload)))
        } catch (cause) {
          throw codecError("payload does not decode as Avro", "decode", cause)
        }
      case "protobuf":
        try {
          const message = this.definition.message.decode(payload)
          return this.definition.message.toObject(message, {
            longs: String,
            enums: String,
            bytes: Array,
            defaults: false,
            oneofs: true
          })
        } catch (cause) {
          throw codecError("payload does not decode as Protobuf", "decode", cause)
        }
      case "jsonSchema": {
        let value: unknown
        try {
          value = JSON.parse(new TextDecoder().decode(payload)) as unknown
        } catch (cause) {
          throw codecError("payload does not parse as JSON", "decode", cause)
        }
        if (!this.definition.validate(value)) {
          throw codecError("payload fails its JSON Schema", "validate")
        }
        return value
      }
    }
  }

  encode(value: unknown): Uint8Array {
    switch (this.definition.kind) {
      case "avro":
        try {
          return Uint8Array.from(this.definition.schema.toBuffer(value))
        } catch (cause) {
          throw codecError("body does not match the Avro schema", "encode", cause)
        }
      case "protobuf": {
        const validation = this.definition.message.verify(value as Record<string, unknown>)
        if (validation !== null) {
          throw codecError(`body does not match the Protobuf schema: ${validation}`, "encode")
        }
        try {
          return this.definition.message
            .encode(this.definition.message.fromObject(value as Record<string, unknown>))
            .finish()
        } catch (cause) {
          throw codecError("Protobuf message encode failed", "encode", cause)
        }
      }
      case "jsonSchema":
        if (!this.definition.validate(value)) {
          throw codecError("body fails its JSON Schema", "validate")
        }
        try {
          return new TextEncoder().encode(JSON.stringify(value))
        } catch (cause) {
          throw codecError("body does not encode as JSON", "encode", cause)
        }
    }
  }

  codec<T>(decodeValue: (value: unknown) => T): Codec<T> {
    return {
      encode: (value) => this.encode(value),
      decode: (payload) => decodeValue(this.decode(payload))
    }
  }
}

File: sdk/src/schema_codecs.rs
use crate::error::LaserError;
use crate::query::{SchemaDef, SchemaSource};
use prost_reflect::{DescriptorPool, DynamicMessage, MessageDescriptor};
use serde::Serialize;
use std::sync::Arc;

/// A registered writer schema compiled for client-side use: encode Avro bodies
/// from serde values and validate payloads against the definition BEFORE
/// publishing, with the same decode semantics LaserData Cloud's projector applies.
/// Without this, the first feedback about a bad schema-first payload is a
/// managed-side warning the producer cannot see.
///
/// Compile once per schema and reuse the result across publishes (parsing is
/// the expensive part):
///
/// ```no_run
/// # use laser_sdk::prelude::*;
/// # use laser_sdk::schema_codecs::CompiledSchema;
/// # use serde::Serialize;
/// # #[derive(Serialize)] struct Order { customer: String, amount: i64 }
/// # async fn run(laser: &Laser, order: Order) -> Result<(), LaserError> {
/// let info = laser.schemas().get(7).await?.expect("schema registered");
/// let compiled = CompiledSchema::compile(&info.schema)?;
/// let orders = laser.topic("orders");
/// orders.publish()
///     .index("customer", &order.customer)
///     .avro(&compiled, 7, &order)?
///     .send().await?;
/// # Ok(()) }
/// ```
#[derive(Clone)]
pub enum CompiledSchema {
    /// A parsed Avro writer schema.
    Avro(apache_avro::Schema),
    /// A resolved Protobuf message descriptor.
    Protobuf(MessageDescriptor),
    /// A compiled JSON Schema validator (draft 2020-12) for the
    /// self-describing codecs.
    Json(Arc<jsonschema::Validator>),
}

impl std::fmt::Debug for CompiledSchema {
    fn fmt(&self, formatter: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            Self::Avro(schema) => formatter.debug_tuple("Avro").field(schema).finish(),
            Self::Protobuf(descriptor) => {
                formatter.debug_tuple("Protobuf").field(descriptor).finish()
            }
            Self::Json(_) => formatter.debug_tuple("Json").finish(),
        }
    }
}

impl CompiledSchema {
    /// Parse a registered [`SchemaDef`] into its compiled form. Fails with
    /// [`LaserError::Invalid`] when the definition does not parse (malformed
    /// Avro JSON, undecodable descriptor set, or a missing message type) -
    /// the same definitions LaserData Cloud would skip with a warning.
    pub fn compile(def: &SchemaDef) -> Result<Self, LaserError> {
        match &def.source {
            SchemaSource::Avro { schema } => {
                let parsed = apache_avro::Schema::parse_str(schema).map_err(|error| {
                    LaserError::Invalid(format!("schema {}: unparseable Avro: {error}", def.id))
                })?;
                Ok(Self::Avro(parsed))
            }
            SchemaSource::Protobuf {
                descriptor_set,
                message_type,
            } => {
                let pool = DescriptorPool::decode(descriptor_set.as_slice()).map_err(|error| {
                    LaserError::Invalid(format!(
                        "schema {}: undecodable Protobuf descriptor set: {error}",
                        def.id
                    ))
                })?;
                let descriptor = pool.get_message_by_name(message_type).ok_or_else(|| {
                    LaserError::Invalid(format!(
                        "schema {}: descriptor set has no message `{message_type}`",
                        def.id
                    ))
                })?;
                Ok(Self::Protobuf(descriptor))
            }
            SchemaSource::JsonSchema { schema } => {
                let value: serde_json::Value = serde_json::from_str(schema).map_err(|error| {
                    LaserError::Invalid(format!(
                        "schema {}: JSON Schema is not valid JSON: {error}",
                        def.id
                    ))
                })?;
                let validator = jsonschema::validator_for(&value).map_err(|error| {
                    LaserError::Invalid(format!(
                        "schema {}: JSON Schema does not compile: {error}",
                        def.id
                    ))
                })?;
                Ok(Self::Json(Arc::new(validator)))
            }
            _ => Err(LaserError::Invalid(format!(
                "schema {}: unknown schema source kind",
                def.id
            ))),
        }
    }

    /// Whether `payload` decodes (Avro, Protobuf) or, for a JSON Schema,
    /// parses as JSON text and passes validation, exactly the check
    /// LaserData Cloud's projector runs before indexing body fields. `false`
    /// means the record would fall back to header-only `agdx.idx.*` extraction. For
    /// non-JSON self-describing payloads (MessagePack, CBOR, BSON), decode
    /// them yourself and use [`CompiledSchema::validate_value`].
    pub fn validate(&self, payload: &[u8]) -> bool {
        self.decode(payload).is_ok()
    }

    /// Validate an already-decoded payload against a JSON Schema, the check
    /// LaserData Cloud runs on a self-describing record stamping `agdx.sid`. Avro and
    /// Protobuf schemas return `false`: stamping their id on a
    /// self-describing record is itself the mismatch LaserData Cloud reports.
    pub fn validate_value(&self, value: &serde_json::Value) -> bool {
        match self {
            Self::Json(validator) => validator.is_valid(value),
            Self::Avro(_) | Self::Protobuf(_) => false,
        }
    }

    /// Decode `payload` under this schema, lowered to a `serde_json::Value` -
    /// the same model LaserData Cloud extracts indexed fields from. Use it to check
    /// which JSON pointers a projection's extraction plan would resolve.
    pub fn decode(&self, payload: &[u8]) -> Result<serde_json::Value, LaserError> {
        match self {
            Self::Avro(schema) => {
                let mut cursor = payload;
                let value =
                    apache_avro::from_avro_datum(schema, &mut cursor, None).map_err(|error| {
                        LaserError::Codec(format!("payload does not decode as Avro: {error}"))
                    })?;
                apache_avro::from_value::<serde_json::Value>(&value).map_err(|error| {
                    LaserError::Codec(format!("Avro value does not lower to JSON: {error}"))
                })
            }
            Self::Protobuf(descriptor) => {
                let message =
                    DynamicMessage::decode(descriptor.clone(), payload).map_err(|error| {
                        LaserError::Codec(format!("payload does not decode as Protobuf: {error}"))
                    })?;
                serde_json::to_value(&message).map_err(|error| {
                    LaserError::Codec(format!("Protobuf message does not lower to JSON: {error}"))
                })
            }
            Self::Json(validator) => {
                let value: serde_json::Value =
                    serde_json::from_slice(payload).map_err(|error| {
                        LaserError::Codec(format!("payload does not parse as JSON: {error}"))
                    })?;
                if !validator.is_valid(&value) {
                    return Err(LaserError::Codec(
                        "payload fails its JSON Schema".to_owned(),
                    ));
                }
                Ok(value)
            }
        }
    }

    /// Encode a serde value as a raw Avro datum (single-object encoding, no
    /// container header), exactly the bytes a producer stamps alongside
    /// `agdx.sid`. Avro schemas only. A Protobuf schema returns
    /// [`LaserError::Invalid`] (encode Protobuf bodies with `prost` and ship
    /// them via `.raw_bytes(bytes, ContentType::Protobuf)`).
    pub fn encode_avro<T: Serialize>(&self, body: &T) -> Result<Vec<u8>, LaserError> {
        let Self::Avro(schema) = self else {
            return Err(LaserError::Invalid(
                "encode_avro requires an Avro schema".to_owned(),
            ));
        };
        let value = apache_avro::to_value(body)
            .map_err(|error| LaserError::Codec(format!("body does not lower to Avro: {error}")))?;
        let resolved = value.resolve(schema).map_err(|error| {
            LaserError::Codec(format!("body does not match the Avro schema: {error}"))
        })?;
        apache_avro::to_avro_datum(schema, resolved)
            .map_err(|error| LaserError::Codec(format!("Avro datum encode failed: {error}")))
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    const ORDER_AVRO_SCHEMA: &str = r#"{
        "type":"record","name":"Order",
        "fields":[
            {"name":"customer","type":"string"},
            {"name":"amount","type":"long"}
        ]
    }"#;

    #[derive(Serialize)]
    struct Order {
        customer: String,
        amount: i64,
    }

    fn avro_def() -> SchemaDef {
        SchemaDef {
            id: 7,
            source: SchemaSource::Avro {
                schema: ORDER_AVRO_SCHEMA.to_owned(),
            },
            name: None,
            version: None,
        }
    }

    #[test]
    fn given_an_avro_def_when_encoded_then_should_decode_and_validate_back() {
        let compiled = CompiledSchema::compile(&avro_def()).expect("schema compiles");
        let order = Order {
            customer: "alice".to_owned(),
            amount: 42,
        };
        let datum = compiled.encode_avro(&order).expect("body encodes");
        assert!(compiled.validate(&datum), "own encoding validates");
        let value = compiled.decode(&datum).expect("datum decodes");
        assert_eq!(
            value.pointer("/customer").and_then(|v| v.as_str()),
            Some("alice")
        );
        assert_eq!(value.pointer("/amount").and_then(|v| v.as_i64()), Some(42));
    }

    #[test]
    fn given_a_mismatched_body_when_encoded_then_should_error() {
        #[derive(Serialize)]
        struct Wrong {
            unrelated: bool,
        }
        let compiled = CompiledSchema::compile(&avro_def()).expect("schema compiles");
        let result = compiled.encode_avro(&Wrong { unrelated: true });
        assert!(matches!(result, Err(LaserError::Codec(_))));
    }

    #[test]
    fn given_garbage_bytes_when_validated_then_should_be_false() {
        let compiled = CompiledSchema::compile(&avro_def()).expect("schema compiles");
        assert!(!compiled.validate(b""));
    }

    #[test]
    fn given_an_unparseable_avro_def_when_compiled_then_should_be_invalid() {
        let def = SchemaDef {
            id: 9,
            source: SchemaSource::Avro {
                schema: "{not valid".to_owned(),
            },
            name: None,
            version: None,
        };
        assert!(matches!(
            CompiledSchema::compile(&def),
            Err(LaserError::Invalid(_))
        ));
    }

    #[test]
    fn given_a_json_schema_def_when_compiled_then_should_validate_and_decode_json() {
        let def = SchemaDef {
            id: 5,
            source: SchemaSource::JsonSchema {
                schema: r#"{
                    "type":"object",
                    "required":["customer","amount"],
                    "properties":{
                        "customer":{"type":"string"},
                        "amount":{"type":"integer","minimum":0}
                    }
                }"#
                .to_owned(),
            },
            name: None,
            version: None,
        };
        let compiled = CompiledSchema::compile(&def).expect("schema compiles");
        assert!(compiled.validate(br#"{"customer":"alice","amount":42}"#));
        assert!(!compiled.validate(br#"{"customer":"alice","amount":"42"}"#));
        assert!(!compiled.validate(b"not json"));
        assert!(compiled.validate_value(&serde_json::json!({"customer":"a","amount":1})));
        assert!(!compiled.validate_value(&serde_json::json!({"amount":1})));
        let value = compiled
            .decode(br#"{"customer":"alice","amount":42}"#)
            .expect("valid payload decodes");
        assert_eq!(
            value.pointer("/customer").and_then(|v| v.as_str()),
            Some("alice")
        );
        assert!(matches!(
            compiled.decode(br#"{"amount":42}"#),
            Err(LaserError::Codec(_))
        ));
        assert!(matches!(
            compiled.encode_avro(&serde_json::json!({})),
            Err(LaserError::Invalid(_))
        ));
    }

    #[test]
    fn given_an_uncompilable_json_schema_def_when_compiled_then_should_be_invalid() {
        let def = SchemaDef {
            id: 6,
            source: SchemaSource::JsonSchema {
                schema: "{not json".to_owned(),
            },
            name: None,
            version: None,
        };
        assert!(matches!(
            CompiledSchema::compile(&def),
            Err(LaserError::Invalid(_))
        ));
    }

    #[test]
    fn given_an_avro_def_when_value_validated_then_should_report_family_mismatch() {
        let compiled = CompiledSchema::compile(&avro_def()).expect("schema compiles");
        assert!(!compiled.validate_value(&serde_json::json!({"customer":"a","amount":1})));
    }

    #[test]
    fn given_a_protobuf_def_when_compiled_then_should_validate_real_messages() {
        use prost::Message;
        let dir = tempfile::tempdir().expect("temp dir");
        let proto = dir.path().join("order.proto");
        std::fs::write(
            &proto,
            "syntax = \"proto3\";\npackage shop;\nmessage Order { string customer = 1; int64 amount = 2; }\n",
        )
        .expect("proto written");
        let descriptor_set = protox::compile([&proto], [dir.path()])
            .expect("proto compiles")
            .encode_to_vec();
        let def = SchemaDef {
            id: 3,
            source: SchemaSource::Protobuf {
                descriptor_set: descriptor_set.clone(),
                message_type: "shop.Order".to_owned(),
            },
            name: None,
            version: None,
        };
        let compiled = CompiledSchema::compile(&def).expect("schema compiles");

        let pool = DescriptorPool::decode(descriptor_set.as_slice()).expect("pool decodes");
        let descriptor = pool.get_message_by_name("shop.Order").expect("message");
        let mut message = DynamicMessage::new(descriptor);
        message.set_field_by_name("customer", prost_reflect::Value::String("alice".into()));
        message.set_field_by_name("amount", prost_reflect::Value::I64(42));
        let payload = message.encode_to_vec();

        assert!(compiled.validate(&payload));
        assert!(!compiled.validate(b"\xff\xff\xff"));
        let value = compiled.decode(&payload).expect("decodes");
        assert_eq!(
            value.pointer("/customer").and_then(|v| v.as_str()),
            Some("alice")
        );
        let encode = compiled.encode_avro(&serde_json::json!({}));
        assert!(matches!(encode, Err(LaserError::Invalid(_))));
    }
}
